version/update.go

```diff
@@ -0,0 +1,184 @@
+package version
+
+import (
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	goversion "github.com/hashicorp/go-version"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	fetchPeriod = 30 * time.Minute
+)
+
+var (
+	versionURL = "https://pkgs.netbird.io/releases/latest/version"
+)
+
+// Update fetch the version info periodically and notify the onUpdateListener in case the UI version or the
+// daemon version are deprecated
+type Update struct {
+	uiVersion     *goversion.Version
+	daemonVersion *goversion.Version
+	lastAvailable *goversion.Version
+	versionsLock  sync.Mutex
+
+	fetchTicker *time.Ticker
+	fetchDone   chan struct{}
+
+	onUpdateListener func()
+	listenerLock     sync.Mutex
+}
+
+// NewUpdate instantiate Update and start to fetch the new version information
+func NewUpdate() *Update {
+	currentVersion, err := goversion.NewVersion(version)
+	if err != nil {
+		currentVersion, _ = goversion.NewVersion("0.0.0")
+	}
+
+	lastAvailable, _ := goversion.NewVersion("0.0.0")
+
+	u := &Update{
+		lastAvailable: lastAvailable,
+		uiVersion:     currentVersion,
+		fetchTicker:   time.NewTicker(fetchPeriod),
+		fetchDone:     make(chan struct{}),
+	}
+	go u.startFetcher()
+	return u
+}
+
+// StopWatch stop the version info fetch loop
+func (u *Update) StopWatch() {
+	u.fetchTicker.Stop()
+
+	select {
+	case u.fetchDone <- struct{}{}:
+	default:
+	}
+}
+
+// SetDaemonVersion update the currently running daemon version. If new version is available it will trigger
+// the onUpdateListener
+func (u *Update) SetDaemonVersion(newVersion string) bool {
+	daemonVersion, err := goversion.NewVersion(newVersion)
+	if err != nil {
+		daemonVersion, _ = goversion.NewVersion("0.0.0")
+	}
+
+	u.versionsLock.Lock()
+	if u.daemonVersion != nil && u.daemonVersion.Equal(daemonVersion) {
```

Comment:
I am not sure if it should be equal or grater than as we are setting it to 0.0.0.0 in case of error. Then override the current value with 0 and next time the current lates would be recognised as new

Reply:
I do not understand well your case.
So for example if the daemon version is default "development" string, we set the version to "0.0.0". In the update check will indicate that new version is available. I think this is the expected behavior.
I use equal instead of grater then because we would like to handle the downgrade events also.